### PR TITLE
Route builder for RDS

### PR DESCRIPTION
**Pull request by nshankar13**
Signed-off-by: nshankar13 <nshankar@microsoft.com>

**Description**: Resolves https://github.com/openservicemesh/osm/issues/4924

**Testing done**:

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change?

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)?

### CODE REVIEW

**Comment by nshankar13** [comment]
@steeling @shashankram few thoughts: 
1) Wasn't sure if "getTrafficPolicy" methods should be in builder or remain in response.go. 
2) Bulk of config building was happening already in route_config.go. Should we merge this into the builder class, or continue to keep them separate? Advantage of having them separate is that builder.go code looks a lot cleaner and more intuitive.

File: pkg/envoy/rds/builder.go
Comment (steeling) [inline pkg/envoy/rds/builder.go:22]: instead of building here, we should be setting all inputs upfront, then call build at the end.

ie: add the TrustDomain as a field, trafficmatches, etc.

File: pkg/envoy/rds/builder.go
Comment (steeling) [inline pkg/envoy/rds/builder.go:75]: let's bring this back to the response.go file

File: pkg/envoy/rds/response.go
Comment (steeling) [inline pkg/envoy/rds/response.go:35]: so for instance here, we may have separate calls:

```go
builder.SetHTTPRouteCOnfigsPerPort(configsPerPort)
.SetProxy(proxy)
.SetTrustDomain(trustDomain)
```

Note that we should not set the cfg, or any other client objects. Instead we should call the client objects here, and set the data on the builder

File: pkg/envoy/rds/response.go
Comment (steeling) [inline pkg/envoy/rds/response.go:63]: Notice that on a lot of these we call GetIngressTrafficPolicy (or inboundTrafficPolicy, outboundTrafficPolicy), but then we only access one field. That means we should probably add a method to just get that one field, so we don't perform the extra work.

No need to tackle in this PR but probably worth a TODO

File: pkg/envoy/rds/builder.go
Comment (steeling) [inline pkg/envoy/rds/builder.go:43]: instead of passing these to the route package's `BuildInboundMeshRouteConfiguration`, let's make BuildInboundMeshRouteConfiguration an unexported method on the builder object.

From the [design doc](https://docs.google.com/document/d/1L_u-XlVU8Fog-uFS3Uf0O2nz8IArQy82cd6Npxm038Y/edit#heading=h.86n6n6h1rqmo), we call out that it's easier when leveraging code navigation to highlight ie: `b.proxy`, and see all of its uses. Whereas when we pass it to another function, we lose that ability.

**Review by steeling — CHANGES_REQUESTED** [review]

File: pkg/envoy/rds/builder.go
Comment (nshankar13) [reply pkg/envoy/rds/builder.go:43]: Makes sense, this is sort of what I was alluding to in my top comment (https://github.com/openservicemesh/osm/pull/4973#issuecomment-1208494343) about merging / borrowing some of the methods in route_config.go and implementing that in builder.go instead. 

Let's discuss more over a call, I think that would be easier.

**Comment by nshankar13** [comment]
Putting this on hold temporarily until globalRateLimiting has been merged, but plan on updating this with by moving some of the methods in route_config (BuildInboundMeshConfiguration) etc, to builder.go, and editing the tests accordingly.

File: pkg/envoy/rds/builder.go
Comment (steeling) [inline pkg/envoy/rds/builder.go:27]: i think the builder and the route package should be joined together. It would make sense to me to move this file into pkg/envoy/rds/route

**Review by steeling — APPROVED** [review]

**Comment by steeling** [comment]
approved! please make sure tests pass before submitting :)

**Review by jsturtevant — APPROVED** [review]
lgtm, pending linter updates

File: pkg/envoy/rds/response.go
Comment (shashankram) [inline pkg/envoy/rds/response.go:75]: `return nil, err` here.

**Review by shashankram — COMMENTED** [review]
Could you collapse pkg/envoy/rds/route/* under pkg/envoy/rds for consistency with other verticals (cds, lds, etc.).

File: pkg/envoy/rds/builder.go
Comment (nshankar13) [reply pkg/envoy/rds/builder.go:27]: As per Shashank's comment I will be removing `/route` and moving everything to `/rds`

File: pkg/envoy/rds/route_config_test.go
Comment (shashankram) [inline pkg/envoy/rds/route_config_test.go:1]: Was this file moved using `git mv`? Curious because this file diff doesn't appear the same as `route_config.go`:
image

Comment (nshankar13) [reply pkg/envoy/rds/route_config_test.go:1]: I moved it and removed the original `route` folder through VS code directly, but I also had to make changes to `route_config.go` because I moved a few of the methods there to `builder.go` and also needed to make some of the methods in `route_config.go` exportable to call in builder.

Comment (shashankram) [reply pkg/envoy/rds/route_config_test.go:1]: Sure, but could you try to confirm if file history can be preserved using `git mv` versus removing the file directly on your IDE?

Comment (nshankar13) [reply pkg/envoy/rds/route_config_test.go:1]: Still not working with `git mv` not sure why it considers route_config_test a new file altogether but preserves version history for the others. If I recreate the /route folder and add route_config_test back there, git lists it as a modification, but if I use `git mv` and move it to /rds and delete the /route folder, then it thinks it's a new file and doesn't preserve the file history.

Comment (shashankram) [reply pkg/envoy/rds/route_config_test.go:1]: Based on discussing this offline, it seems the file was moved after being edited, instead of being moved first. This could cause the file history to be unexpected. @nshankar13 will try to fix this.

Comment (nshankar13) [reply pkg/envoy/rds/route_config_test.go:1]: Tried resetting, moving the file with `git mv`, and then applying edits on top of the moved files. However, git still thinks `rds/builder_test.go` was previously `rds/route/route_config_test.go`, and that `rds/route_config_test.go` is now a new file.

Comment (nshankar13) [reply pkg/envoy/rds/route_config_test.go:1]: `pkg/envoy/rds/route/route_config_test.go → pkg/envoy/rds/builder_test.go`

Comment (nshankar13) [reply pkg/envoy/rds/route_config_test.go:1]: This might have to do with how `git` detects renames: https://chelseatroy.com/2020/05/09/question-how-does-git-detect-renames/

File: pkg/envoy/rds/builder.go
Comment (keithmattix) [inline pkg/envoy/rds/builder.go:14]: nit: if these are prefixes, that should be reflected in the variable name

**Review by keithmattix — APPROVED** [review]
One naming nit, but it's not blocking. Great work!

**Pull request merged**
